allauth/socialaccount/providers/dataporten/views.py

```diff
@@ -0,0 +1,75 @@
+import requests
+
+from django.core.exceptions import SuspiciousOperation
+
+from allauth.socialaccount.providers.oauth2.views import (
+    OAuth2Adapter,
+    OAuth2CallbackView,
+    OAuth2LoginView,
+)
+
+from .provider import DataportenProvider
+
+
+class DataportenAdapter(OAuth2Adapter):
+    provider_id = DataportenProvider.id
+    access_token_url = 'https://auth.dataporten.no/oauth/token'
+    authorize_url = 'https://auth.dataporten.no/oauth/authorization'
+    profile_url = 'https://auth.dataporten.no/userinfo'
+    groups_url = 'https://groups-api.dataporten.no/groups/'
+
+    def complete_login(self, request, app, token, **kwargs):
+        '''
+        Arguments:
+            request - The get request to the callback URL
+                        /accounts/dataporten/login/callback.
+            app - The corresponding SocialApp model instance
+            token - A token object with access token given in token.token
+        Returns:
+            Should return a dict with user information intended for parsing
+            by the methods of the DataportenProvider view, i.e.
+            extract_uid(), extract_extra_data(), and extract_common_fields()
+        '''
+        # The athentication header
+        headers = {'Authorization': 'Bearer ' + token.token}
+
+        # Userinfo endpoint, for documentation see:
+        # https://docs.dataporten.no/docs/oauth-authentication/
+        userinfo_response = requests.get(
+            self.profile_url,
+            headers=headers,
+        )
+        # Raise exception for 4xx and 5xx response codes
+        userinfo_response.raise_for_status()
+
+        # Groups endpoint, for documentation see:
+        # https://docs.dataporten.no/docs/groups/
+        groups_data = requests.get(
+            self.groups_url + 'me/groups',
+            headers=headers,
+        )
+        # Raise exception for 4xx and 5xx response codes
+        groups_data.raise_for_status()
+
+        # The endpoint returns json-data and it needs to be decoded
+        extra_data = userinfo_response.json()['user']
+        extra_data['groups'] = groups_data.json()
+
+        # Finally test that the audience property matches the client id
+        # for validification reasons, as instructed by the Dataporten docs
+        # if the userinfo-response is used for authentication
+        if userinfo_response.json()['audience'] != app.client_id:
```

Comment:
No, this is fine... a specific one would make sense if it somehow could be handled in a special way, but that's not the case.